lib/pry/input_completer.rb

```diff
@@ -239,4 +227,39 @@ def build_path(input)
     end
     return path, input
   end
+
+  private
+
+  def self.fetch_obj_space_collection
```

Comment:
i don't understand that argument.
an instance of Pry::InputCompleter already belongs to an instance of Pry, we just re-initialise it every time Pry accepts input. Avoiding global state is always a good thing, and the only reason it has to be global is because we throw away the instance of Pry::InputCompleter at each eval tick. we could initialise _one_ instance of Pry::InputCompleter, store the method cache on that instance, and avoid the lock altogether. locks should be a last resort, and global state should be avoided.